Bl/repo page

beyang opened the pull request:
## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.02% (+0.50 kb) | 0.04% (+6.00 kb) | 0.05% (+5.50 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/bl/repo-page/compare-report.html;tab=live_object) for a full comparison between the commits 3a8bfec and eddcd34 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

bobheadxi commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,322 @@
+    readmeURL,
+    location,
+}) => {
+    const fileRef = useRef<HTMLDivElement>(null)
+    const containerRef = useRef<HTMLDivElement>(null)
+    const isNotCutoff =
+        fileRef.current &&
+        containerRef.current &&
+        fileRef.current.clientHeight > 0 &&
+        containerRef.current.clientHeight >= fileRef.current.clientHeight
```
This approach seems to cause a "flicker" - for example, when viewing a directory like:

```
repo:^github\.com/sourcegraph/sourcegraph$ file:^cmd
```

The file when render, the "More..." button will render briefly, and then it will disappear. I think it's because this particular file ends up with a height that's almost exactly the height of the container?

bobheadxi replied:
It looks like it behaves a bit better if we add a bit of a padding to the condition:

```suggestion
    const fileRef = useRef<HTMLDivElement>(null)
    const containerRef = useRef<HTMLDivElement>(null)
    const isNotCutoff =
        fileRef.current &&
        containerRef.current &&
        fileRef.current.clientHeight > 0 &&
        containerRef.current.clientHeight >= fileRef.current.clientHeight - 4
```

bobheadxi commented on client/web/src/repo/tree/TreePageContent.tsx:
```diff
@@ -82,178 +238,382 @@ export const TreePageContent: React.FunctionComponent<React.PropsWithChildren<Tr
+                <PersonLink userClassName="font-weight-bold" person={node.person} />
+            </div>
+            <div className={contributorsStyles.commits}>
+                <div className={contributorsStyles.commit}>
+                    {commit && (
+                        <>
+                            {/* <Timestamp date={commit.author.date} />:{' '} */}
+                            <Tooltip content="Most recent commit by contributor" placement="bottom">
+                                <Link to={commit.url} className="repository-contributor-node__commit-subject">
+                                    {commit.subject}
+                                </Link>
+                            </Tooltip>
+                        </>
+                    )}
+                </div>
```
IMO this most recent commits isn't necessary - in smaller and less active subdirectories, we just get a more or less duplicated list:

<img width="1230" alt="image" src="https://user-images.githubusercontent.com/23356519/206830336-88ea0fec-1bcb-4ba2-9f48-65c59b1eb812.png">

bobheadxi replied:
```suggestion
```

With it removed:

<img width="1071" alt="image" src="https://user-images.githubusercontent.com/23356519/206830443-42c92eae-0df6-4c44-b9ab-3fc35791e5d8.png">


I think it looks easier to read on more busy directories too:

<img width="1240" alt="image" src="https://user-images.githubusercontent.com/23356519/206830485-179eff69-38ba-4085-94b0-376755504c0c.png">

bobheadxi commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,322 @@
+                        : classNames(styles.readmeMore, styles.readmeMoreVisible)
+                }
+            >
+                <Link to={readmeURL}>More...</Link>
```
It opens a new page view so IMO:

```suggestion
                <Link to={readmeURL}>View full file</Link>
```

bobheadxi commented on client/web/src/repo/tree/TreePage.module.scss:
```diff
@@ -41,3 +59,102 @@
+}
+
+.readme-more {
+    margin-top: 0;
```
IMO this needs a touch of padding:

```suggestion
    margin-top: 0.5rem;
```

bobheadxi reviewed (commented):
This looks great! Made a first pass review, but probably want a review from the folks over at #dev-frontend as well 😁 

Another minor nit: I think you are just using existing components, but the style differences between `Commits` and `Contributors` when they are side by side is a bit jarring (bolded text, the footer text styles, etc):

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/23356519/206830712-11a07371-9798-47f2-a824-31b0605c140a.png">

beyang commented on docker-images/prometheus/Dockerfile:
```diff
@@ -77,7 +77,6 @@ COPY ./alertmanager.sh /alertmanager.sh
 
 # Copy config
 COPY --from=monitoring_builder /generated/prometheus/* /sg_config_prometheus/
-COPY config/*_rules.yml /sg_config_prometheus/
```
It looks like there haven't been any `*_rules.yml` files in that directory since https://github.com/sourcegraph/sourcegraph/commit/aa05d0bd2497d0e5f6a02e27f0edc3410817f950 and https://github.com/sourcegraph/sourcegraph/commit/834da914c860bc0bb30cb7a3a7e5e2b8061f6951. It broke for me when I ran it locally—are we relying on this somewhere else?

bobheadxi replied:
They are generated when the image gets built, and are not committed: 

```
➜ sg monitoring generate
➜ ls docker-images/prometheus/config
alertmanager.yml                          executor_alert_rules.yml                  prometheus_alert_rules.yml                syntect_server_alert_rules.yml
codeintel_autoindexing_alert_rules.yml    frontend_alert_rules.yml                  prometheus_targets.yml                    telemetry_alert_rules.yml
codeintel_codenav_alert_rules.yml         github_proxy_alert_rules.yml              redis_alert_rules.yml                     worker_alert_rules.yml
codeintel_policies_alert_rules.yml        gitserver_alert_rules.yml                 repo_updater_alert_rules.yml              zoekt_alert_rules.yml
codeintel_ranking_alert_rules.yml         postgres_alert_rules.yml                  searcher_alert_rules.yml
codeintel_uploads_alert_rules.yml         precise_code_intel_worker_alert_rules.yml src_custom_rules.yml
containers_alert_rules.yml                prometheus.yml                            symbols_alert_rules.yml
```

bobheadxi commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,322 @@
+                                >
+                                    <ParentSize>
+                                        {({ width }) => (
+                                            <StackedMeter
+                                                width={width}
+                                                height={10}
+                                                viewMinMax={[0, maxLinesChanged]}
+                                                data={[
+                                                    {
+                                                        name: 'deleted',
+                                                        value: diffStatsByPath[entry.name].deleted,
+                                                        className: styles.diffStatDeleted,
+                                                    },
+                                                    {
+                                                        name: 'added',
+                                                        value: diffStatsByPath[entry.name].added,
+                                                        className: styles.diffStatAdded,
+                                                    },
+                                                ]}
+                                                getDatumValue={getDatumValue}
+                                                getDatumName={getDatumName}
+                                                getDatumClassName={getDatumClassName}
+                                                minBarWidth={10}
+                                                className={styles.barSvg}
+                                                barRadius={5}
+                                                rightToLeft={true}
+                                            />
```
What these bars represent is hard to make sense of in the UI - I think we needed the added/deleted bars to have descriptions/summaries on hover

vovakulikov replied:
I also noticed that this chart doesn't work really well with a screen reader. I think we should add there tooltip UI or bar labels as @bobheadxi suggested

beyang replied:
They have hover text currently. Is there a UI element I should use?

bobheadxi replied:
Ah, I see it - the delay on the hover is quite slow. It also says `+X/-Y lines`, which doesn't feel descriptive enough - I would expect `X lines added, Y lines deleted in past Z days`

beyang replied:
Switched to using `<Tooltip>`

vovakulikov commented on client/branded/src/global-styles/colors.scss:
```diff
@@ -205,6 +205,8 @@ $theme-colors: (
     --code-line-highlight-color: var(--color-bg-2);
     --diff-add-bg: #eeffec;
     --diff-remove-bg: #ffecec;
+    --diff-add-fg: #93dc8b;
```
I'm just curious, what does the suffix "fg" mean?

beyang replied:
"foreground" (as opposed to "bg" meaning "background")

vovakulikov commented on client/web/src/repo/tree/TreePage.module.scss:
```diff
@@ -41,3 +59,102 @@
+    overflow: hidden;
+    position: relative;
+    margin-top: 1rem;
+    margin-left: -0.25rem;
```
Can we just set padding-left: 0 to the readme container?

vovakulikov commented on client/web/src/repo/tree/TreePage.module.scss:
```diff
@@ -18,8 +20,7 @@
 }
 
 .git-commit-node {
```
General comment about styles. I see that this styles file is used in many child UI of the repo page, I would suggest either 
- Structure UI component in a way that it would be possible to pass all classes at one place (top level page) for example 
```tsx
<Page className={styles.page}>
    <RepoPageNavigation className={styles.pageNavigation} 
    <RepoPageReadme className={styles.pageReadme} 
    // ...
</Page>
```
- Split these top-level styles file into small and put them closer to the places where we use these styles 
```
- TreePage.tsx
- TreePage.modules.scss
- /components
  - TreePageNavigation.tsx
  - TreePageNavigation.modules.scss
  ...
```

vovakulikov commented on client/web/src/repo/tree/TreePage.module.scss:
```diff
@@ -30,6 +31,23 @@
     }
 }
 
+.file-stat-selector {
```
I think this class isn't used anywhere.

vovakulikov commented on client/web/src/repo/tree/TreePage.module.scss:
```diff
@@ -41,3 +59,102 @@
+    width: 100%;
+    height: 2.5rem;
+    transition: ease-in;
+    background-image: linear-gradient(transparent, 75%, var(--code-bg));
```
I haven't seen this syntax of `line-gradient` before. I think the latest valid spec provides the following syntax 
`([[<angle> | to? <side-or-corner>] ,]? <color-stop> [, <color-stop>]+)`

So I think you want to have something like this.

```CSS 
 background: linear-gradient(to bottom, transparent, var(--code-bg));
```

Also, note that in the previous version of Safari, we had [problems with a transparent gradient](https://css-tricks.com/thing-know-gradients-transparent-black/). It may be worth using strict colours instead of the `transparent` keyword in order to avoid half black gradient. I checked the new version of Safari and it works fine with `transparent`

vovakulikov commented on client/web/src/repo/tree/TreePage.module.scss:
```diff
@@ -41,3 +59,102 @@
+.readme-more {
+    margin-top: 0.5rem;
+    margin-bottom: 1rem;
+    visibility: hidden;
```
Even if this element has `visibility: hidden;` it's still possible that this element will still be pronounced by a screen reader. 

```
PRODUCT                         DISPLAY: NONE       VISIBILITY: HIDDEN
Hal version 5.20                Does not read       Reads
IBM Home Page Reader 3.02       Does not read       Does not read
Jaws (4.02, 4.50, 5.0 beta)     Reads               Reads
OutSpoken 9                     Does not read       Does not read
Window-Eyes 4.2                 Does not read       Does not read
```

Also, it's a good practice to keep our DOM with a minimal number of elements. I would suggest not rendering this element when we don't need it instead of hiding it with CSS rules.

vovakulikov commented on client/web/src/repo/tree/TreePageContent.tsx:
```diff
@@ -27,42 +28,197 @@ import {
+        })
+    )
 
 interface TreePageContentProps extends ExtensionsControllerProps, ThemeProps, TelemetryProps, PlatformContextProps {
```
I like the new UI, but I think the old layout works better because it introduces more scrolling. I made a quick demo of a possible layout for this page. Similar to what GH has on their repo page.

Not sure, but maybe moving redeame below the file tree section also would make sense.

https://user-images.githubusercontent.com/18492575/206862327-4fe1fd30-9202-4045-99ed-d8e0e2f321b6.mp4

vovakulikov replied:
@beyang I see that you decided to put Commits info below the file structure in the first column, I think I agree with it because in the commits section we have too many info to fit in a narrow aside block.

But a few things that I've noticed there
- The layout implementation seems a bit complicated (too many nested elements and bootstrap col/row styles) 
<img width="555" alt="Screenshot 2023-01-03 at 14 30 00" src="https://user-images.githubusercontent.com/18492575/210409972-dbf75b2a-8011-4559-aa18-a6c2fd41f7e2.png">
- The gap between file structure block and contribution seems a bit bigger than it should be

vovakulikov replied:
I think we could use CSS grid layout to simplify this UI 
Here is a minimal demo with similar UI structure https://jsbin.com/cakefuwuca/edit?html,css,output

vovakulikov commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,322 @@
+}) => {
+    const fileRef = useRef<HTMLDivElement>(null)
+    const containerRef = useRef<HTMLDivElement>(null)
+    const isNotCutoff =
```
I think it's a good practice to have variables without NOT part, not prefix makes this logic hard to unwrap in the reader call IMO

vovakulikov commented on client/wildcard/src/components/Charts/components/stacked-meter/StackedMeter.tsx:
```diff
@@ -0,0 +1,107 @@
+    children,
+    rightToLeft,
+    minBarWidth,
+    barRadius,
```
In order to enforce the standard appereance of StackedMeter, I would like to not having non-vital visual props in public API

vovakulikov commented on client/wildcard/src/components/Charts/components/stacked-meter/StackedMeter.tsx:
```diff
@@ -0,0 +1,107 @@
+                            left={isFirstBar}
+                            right={isLastBar}
+                        >
+                            {children && (({ path }) => children({ path, datum: stackedDatum.datum }))}
```
Do we have a case for children's prop?

vovakulikov commented on client/wildcard/src/components/Charts/components/stacked-meter/index.ts:
```diff
@@ -0,0 +1 @@
+export { StackedMeter as StackedMeter } from './StackedMeter'
```
I think we don't need to have as reexport here 
```
port { StackedMeter } from './StackedMeter'
``` 

works just fine

beyang replied:
```suggestion
export { StackedMeter } from './StackedMeter'
```

vovakulikov reviewed (commented):
Nice work! I left a few comments there, let me know if you need any help with FE part of this PR 😉

rrhyne commented:
Nice! Personally, I'd say remove the duplicative files panel. I know a few other devs feel the same way, but people are scared to remove it. There is no data I am aware of on its use.

rrhyne reviewed (approved):
Nice! Personally, I'd say remove the duplicative files panel. I know a few other devs feel the same way, but people are scared to remove it. There is no data I am aware of on its use.

beyang commented on client/web/src/repo/tree/TreePageContent.tsx:
```diff
@@ -82,178 +238,382 @@ export const TreePageContent: React.FunctionComponent<React.PropsWithChildren<Tr
+                            <CardHeader className={styles.cardColHeaderWrapper}>
+                                {tree.isRoot ? <Link to={`${tree.url}/-/commits`}>Commits</Link> : 'Commits'}
+                            </CardHeader>
+                            <FilteredConnection<
```
I don't yet see `useConnection` in `main`. Candidate for a follow-up PR?

beyang commented:
@valerybugakov @bobheadxi addressed your comments, PTAL

https://user-images.githubusercontent.com/1646931/209782750-a32d098e-cfd4-4ee4-84e9-ceb426fdf35d.mp4

vovakulikov commented:
@beyang Sorry, I haven't noticed that in the first round. With new repo action/setting layout isn't straightforward enough, I may not be familiar with all this icon coding there. 

<img width="368" alt="Screenshot 2023-01-03 at 13 38 49" src="https://user-images.githubusercontent.com/18492575/210400952-59950187-3b38-444b-9634-df4f1974866a.png">

I would suggest to provide text description at least in tooltip on icon hover to make it more clear.

vovakulikov commented on client/wildcard/src/components/Charts/components/stacked-meter/StackedMeter.story.tsx:
```diff
@@ -0,0 +1,71 @@
+                    getDatumValue={getDatumValue}
+                    rightToLeft={true}
+                >
+                    {({ path, datum }) => (
```
I think we don't need to have any children here

vovakulikov commented on client/wildcard/src/components/Charts/components/stacked-meter/StackedMeter.tsx:
```diff
@@ -0,0 +1,102 @@
+        range: [0, width],
+    })
+
+    const adjustedGetDatumValue: (datum: Datum) => number = minBarWidth
```
Let's leave a comment here about `number = minBarWidth` case

vovakulikov commented on client/wildcard/src/components/Charts/components/stacked-meter/StackedMeter.tsx:
```diff
@@ -0,0 +1,102 @@
+    getDatumClassName?: (datum: Datum) => string
+    className?: string
+    rightToLeft?: boolean
+    children?: ({ path, datum }: { path: string; datum: Datum }) => React.ReactNode
```
We probably don't need this anymore.

vovakulikov commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,325 @@
+}
+const formatNumber = (value: number): string => Intl.NumberFormat('en', { notation: 'compact' }).format(value)
+
+export const DiffMeter: React.FunctionComponent<{
```
Should we delete this since you added a designated SVG chart in the wildcard for the meter UI?

vovakulikov commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,325 @@
+                                    <div className={styles.meterContainer}>
+                                        <ParentSize>
+                                            {({ width }) => (
+                                                <div>
```
Why do we need additional wrapper around chart here? could we delete this?

vovakulikov commented:
@beyang I think I found a problem with readme cut off logic. (Check out the video below)

Step to reproduce
- Go to the `sourcegraph/sourcegraph` repo page
- Go to the `/client` directory (check that you can see view more in the readme section)
- Go to the any folder in the `/client` dir
- Go back to the `./client` directory
- See that the view full readme is not rendered

https://user-images.githubusercontent.com/18492575/210407155-2184f001-fbfa-4a7d-98f6-88b2cbe255eb.mov

vovakulikov commented on client/web/src/repo/tree/TreePagePanels.tsx:
```diff
@@ -0,0 +1,325 @@
+}) => {
+    const fileRef = useRef<HTMLDivElement>(null)
+    const containerRef = useRef<HTMLDivElement>(null)
+    const shouldHideReadmeTail = !(
```
I think the problem with showing view the full readme button is hiding here. The problem here that when React set container and file ref it doesn't trigger react render call, so `shouldHideReadmeTail` won't be recalculated the whole `ReadmePreviewCard` won't be called (unless something higher on the tree will be triggered and therefore will trigger `ReadmePreviewCard` again. 

So I would suggest to use useState here instead of useRef if render can affect other element rendering. 

```
[fileElement, setFileElement] = useState<Element|null>(null)
...
ref={setFileElement}
...
```

beyang replied:
nice catch, TIL!

bobheadxi reviewed (commented):
On top of Vova's comments, I also noticed that the `Show full README` button flickers in and out when the file is not large enough to warrant it - we should only render the button _after_ we determine the file is too large:

https://user-images.githubusercontent.com/23356519/210463759-13b6fc1d-92dc-4f4e-8e22-1d7e46c24772.mp4

Otherwise this is looking great!

bobheadxi reviewed (approved):
LGTM, pending review from @vovakulikov and merge conflict fixes :)

Some final ideas:

- it would be nice for the "activity bar" to be clickable to a diff/commit search scoped to the directory
- footer text on commit history for consistency (maybe `Commits since X`)

<img width="1205" alt="Screenshot 2023-01-04 at 2 06 09 PM" src="https://user-images.githubusercontent.com/23356519/210659057-20648c23-fb68-46a5-9844-54d970aa01a8.png">

vovakulikov commented:
@beyang I take a look at this once again and I think there is still a problem with view full readme button. See this comment https://github.com/sourcegraph/sourcegraph/pull/45390#issuecomment-1370027517

Also comments there are some non blocking comments (but if you have a moment to take a look, some of them are very easy to fix/resolve)
- [Additional wrapper element](https://github.com/sourcegraph/sourcegraph/pull/45390#discussion_r1060787856)
- [Old diff meter UI](https://github.com/sourcegraph/sourcegraph/pull/45390#discussion_r1060787355)
- [Old meter chart API](https://github.com/sourcegraph/sourcegraph/pull/45390#discussion_r1060781492)

beyang commented:
@valerybugakov What I'm observing happening with the "View README" button is this:

It's fine on new page load.

It appears to flicker when you navigate to a page with a short (fully visible) README from a page with a long (partially hidden) README. That is because there is a slight delay between when React re-renders the UI and when it updates the ref.

We could try to get around this, but I don't think it's worth it, because:
* The navigation is fast enough when clicking into a subdirectory that the user likely won't notice it unless they're looking specifically for it (your eye focus is on what you just clicked)
* Implementing logic to immediately hide the button after clicking something on the page feels like overkill

I've addressed your other comments.

I think the new page is a significant improvement over the existing page, so I'd advocate for merging this, and any additional minor improvements I can follow up with later.

vovakulikov commented:
@beyang I agree that this PR brings many benefits and I don't want to block this. 

>It appears to flicker when you navigate to a page with a short (fully visible) README from a page with a long (partially hidden) README. That is because there is a slight delay between when React re-renders the UI and when it updates the ref.

The problem that I described here https://github.com/sourcegraph/sourcegraph/pull/45390#issuecomment-1370027517 isn't about just flickering, it shows that as you navigate through different folders with different readme content the full readme button doesn't show up when it's needed. 

I fixed this and a few other other visual problems in the last two commits, feel free to revert or change there anything
